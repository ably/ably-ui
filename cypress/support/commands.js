Cypress.Commands.add(
  "compareComponents",
  ({
    reactComponentUrl,
    viewComponentUrl,
    beforeScreenshot = () => {},
    selector,
    componentName = Cypress.mocha.getRunner().suite.ctx.test.parent.title,
  }) => {
    let reactScreenshotPath = null;
    let vwScreenshotPath = null;

    cy.exec(`mkdir -p cypress/screenshots/component-parity/results`, {
      log: false,
    });

    const diffPath = `cypress/screenshots/component-parity/results/${componentName}.png`;

    // If you are wondering why this isn't Promise.all or async/await have a look at
    // https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

    cy.visit(reactComponentUrl);

    beforeScreenshot();

    cy.get(selector)
      .screenshot(`${componentName}.react`, {
        onAfterScreenshot: (_, props) => (reactScreenshotPath = props.path),
      })
      .then(() => {
        cy.visit(viewComponentUrl);
        beforeScreenshot();

        cy.get(selector)
          .screenshot(`${componentName}.vw`, {
            onAfterScreenshot: (_, props) => (vwScreenshotPath = props.path),
          })
          .then(() => {
            cy.exec(
              `cypress/support/diff.js ` +
                `--pathA="${reactScreenshotPath}" ` +
                `--pathB="${vwScreenshotPath}" ` +
                `--target="${diffPath}" ` +
                `--threshold="0.005" `
            ).then((result) => {
              cy.exec(`rm "${reactScreenshotPath}" "${vwScreenshotPath}"`, {
                log: false,
              });

              assert.isTrue(result.stdout === "Yay", "Screenshots match");
            });
          });
      });
  }
);
