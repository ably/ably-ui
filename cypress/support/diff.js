#! /usr/bin/env node

// Used by the compareComponents command

const Blink = require("blink-diff");
const argv = require("yargs").argv;

const diff = new Blink({
  imageAPath: argv.pathA,
  imageBPath: argv.pathB,
  imageOutputPath: argv.target,
  threshold: 0.05,
});

diff.run((error, result) => {
  // We read stdout/stderr - this script is executed interactively by cypress
  if (error) {
    console.error(error);
  } else {
    console.log(diff.hasPassed(result.code) ? "Yay" : "Nay");
  }
});
