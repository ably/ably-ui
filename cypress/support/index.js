// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import "cypress-plugin-snapshots/commands";
import "./commands";

// Cypress has a bug that breaks snapshot tests when running the whole suite
// Follow https://github.com/cypress-io/cypress/issues/3090
// Recommended workaround below:
export const fixSnapshotSpec = (filename) => () => {
  const path = require("path");
  // const relative = filename.substr(1); // removes leading "/"
  const projectRoot = Cypress.config("projectRoot");
  const absolute = path.join(projectRoot, filename);

  Cypress.spec = {
    absolute,
    name: path.basename(filename),
    filename,
  };
};
