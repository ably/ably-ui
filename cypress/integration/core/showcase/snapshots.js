import { fixSnapshotSpec } from "../../../support";

beforeEach(fixSnapshotSpec(__filename));

describe("Showcase", () => {
  beforeEach(() => {
    cy.viewport("macbook-16");
    cy.visit("/components/showcase?framework=vw");
  });

  it("the button is inactive without Hubspot widget", () => {
    cy.get("[data-id='showcase']").first().toMatchSnapshot();
  });
});
