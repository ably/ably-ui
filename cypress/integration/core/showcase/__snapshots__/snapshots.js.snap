exports[`Showcase > the button is inactive without Hubspot widget #0`] = `
<section
  aria-roledescription="carousel"
  aria-label="Ably customers showcase"
  aria-live="off"
  class="font-sans antialiased rounded-sm shadow-container-subtle"
  data-id="showcase"
>
  <div class="ui-showcase-overflow-container">
    <ol class="ui-showcase-logo-wrapper" data-id="showcase-controls">
      <li class="ui-showcase-logo">
        <button
          class="focus:outline-gui-focus border-none p-0 transition-filter"
          type="button"
          data-id="showcase-control"
          aria-controls="showcase-slides-gi0jchn2"
        >
          <img
            alt="Hubspot logo"
            class="ui-showcase-logo-img filter-none hover:filter-none"
            srcset="
              /assets/ably_ui/core/images/cust-logo-hubspot-col-pos@2x-2a3c7f0277ef4b9c80076e0cd418f4097c82d707a02f0a70c816abd75f6fbcec.png 2x
            "
            src="/assets/ably_ui/core/images/cust-logo-hubspot-col-pos-9e6e33d4e80078537936558cee3dbea02561d8f739877a25eb875678a98218ea.png"
          />
        </button>
      </li>
      <li class="ui-showcase-logo">
        <button
          class="focus:outline-gui-focus border-none p-0 transition-filter"
          type="button"
          data-id="showcase-control"
          aria-controls="showcase-slides-gi0jchn2"
        >
          <img
            alt="Vitac logo"
            class="ui-showcase-logo-img filter-grayscale hover:filter-none"
            srcset="
              /assets/ably_ui/core/images/cust-logo-vitac-col-pos@2x-3bf5a4c07aa32dbf8469d2c23d9fa2132154b426a9b621202b38af0552f3c15f.png 2x
            "
            src="/assets/ably_ui/core/images/cust-logo-vitac-col-pos-5f4093496720f7447d3d20daed601e1699b697dd3185116ba27e087770b4f65c.png"
          />
        </button>
      </li>
      <li class="ui-showcase-logo">
        <button
          class="focus:outline-gui-focus border-none p-0 transition-filter"
          type="button"
          data-id="showcase-control"
          aria-controls="showcase-slides-gi0jchn2"
        >
          <img
            alt="Split logo"
            class="ui-showcase-logo-img filter-grayscale hover:filter-none"
            srcset="
              /assets/ably_ui/core/images/cust-logo-split-col-pos@2x-450c13c37a6676c7d63e62ea4a238f4366e9da5301a3b07aaadbc683659065e2.png 2x
            "
            src="/assets/ably_ui/core/images/cust-logo-split-col-pos-afc19e4f0cb289b2cde99af2770f339026762ce18107f2849d459a641248ba05.png"
          />
        </button>
      </li>
    </ol>

    <div class="ui-showcase-index-bar">
      <div
        data-id="showcase-index-bar"
        class="transform transition-transform w-1/3 h-4 bg-active-orange relative"
      ></div>
    </div>
  </div>

  <div class="overflow-x-hidden">
    <ol
      class="flex transform transition-transform"
      data-id="showcase-slides"
      id="showcase-slides-gi0jchn2"
    >
      <li
        class="w-full flex-shrink-0 p-24 sm:p-32 xl:p-40 bg-white"
        role="group"
        aria-roledescription="slide"
        aria-label="1 of 3"
      >
        <svg class="block mb-16 w-24 h-20">
          <use xlink:href="#sprite-quote"></use>
        </svg>

        <p class="text-quote font-light mb-24">
          Ably’s realtime infrastructure layer seamlessly supports HubSpot's
          entire realtime needs and helps us meet technical, business, and
          product development requirements.
        </p>

        <div class="flex items-center">
          <div
            class="flex-shrink-0 flex-grow-0 mr-16 overflow-hidden rounded-full flex align-center justify-center w-48 h-48"
            style="border-radius: 50%"
          >
            <img
              alt="Picture of Max Freiert, Product Group Lead / HubSpot"
              class="w-48 h-48 filter-grayscale"
              src="/assets/ably_ui/core/images/cust-photo-hubspot-max-freiert-2471d9aab5e31e145c333d7bc70730cf81287dd98ad12a9774729690bd1c9d84.jpg"
            />
          </div>
          <div>
            <p class="text-p2 font-medium">Max Freiert</p>
            <p class="text-p3 font-light">Product Group Lead / HubSpot</p>
          </div>
        </div>
      </li>
      <li
        class="w-full flex-shrink-0 p-24 sm:p-32 xl:p-40 bg-white"
        role="group"
        aria-roledescription="slide"
        aria-label="2 of 3"
      >
        <svg class="block mb-16 w-24 h-20">
          <use xlink:href="#sprite-quote"></use>
        </svg>

        <p class="text-quote font-light mb-24">
          In a high-stakes, highly competitive industry, VITAC sought a provider
          that could operate realtime infrastructure for transporting live data
          to end-users via a complex, multi-hop process. In media accessibility
          environments - where there’s zero margin for error - Ably’s
          infrastructure performs and exceeds expectations.
        </p>

        <div class="flex items-center">
          <div
            class="flex-shrink-0 flex-grow-0 mr-16 overflow-hidden rounded-full flex align-center justify-center w-48 h-48"
            style="border-radius: 50%"
          >
            <img
              alt="Picture of Joe Antonio, Chief Information Officer / VITAC"
              class="w-48 h-48 filter-grayscale"
              src="/assets/ably_ui/core/images/cust-photo-vitac-joe-antonio-7756d510bf7f7ed7bd752d1a9ae204dcaed1f322cc6e054d38097b7cd63e1e28.jpg"
            />
          </div>
          <div>
            <p class="text-p2 font-medium">Joe Antonio</p>
            <p class="text-p3 font-light">Chief Information Officer / VITAC</p>
          </div>
        </div>
      </li>
      <li
        class="w-full flex-shrink-0 p-24 sm:p-32 xl:p-40 bg-white"
        role="group"
        aria-roledescription="slide"
        aria-label="3 of 3"
      >
        <svg class="block mb-16 w-24 h-20">
          <use xlink:href="#sprite-quote"></use>
        </svg>

        <p class="text-quote font-light mb-24">
          You miss so much by not using a platform like Ably. When you need to
          implement a new feature, the capabilities are there, ready to go. Or
          when you need to scale, the capacity is seamlessly available. There’s
          no need to even think about these things. Building on Ably was the
          only logical choice because we managed to bypass a hefty DevOps debt
          and rapidly ship our new streaming capabilities while keeping our
          architecture as simple and reliable as possible.
        </p>

        <div class="flex items-center">
          <div
            class="flex-shrink-0 flex-grow-0 mr-16 overflow-hidden rounded-full flex align-center justify-center w-48 h-48"
            style="border-radius: 50%"
          >
            <img
              alt="Picture of Pato Echagüe, Chief Technical Officer / Split"
              class="w-48 h-48 filter-grayscale"
              src="/assets/ably_ui/core/images/cust-photo-split-pato-echague-09acbdbe033f53bb04c7a60fe169a6c590b386acaa33f49335d80e52db05a604.jpg"
            />
          </div>
          <div>
            <p class="text-p2 font-medium">Pato Echagüe</p>
            <p class="text-p3 font-light">Chief Technical Officer / Split</p>
          </div>
        </div>
      </li>
    </ol>
  </div>
</section>
`;

exports[`Chat button state > react > the button is inactive without Hubspot widget #0`] = `
<button
  type="button"
  class="ui-btn-secondary self-start p-btn mt-16"
  data-id="open-chat-widget"
>
  Start a live chat
</button>
`;

exports[`Chat button state > vw > the button is inactive without Hubspot widget #0`] = `
<button
  type="button"
  class="ui-btn-secondary self-start p-btn mt-16"
  data-id="open-chat-widget"
>
  Start a live chat
</button>
`;

exports[`Footer Snapshot Test > react > test case common to both #0`] = `
<footer class="bg-light-grey font-sans antialiased" data-id="footer">
  <div
    class="max-w-screen-xl mx-auto p-24 md:px-32 lg:p-32 xl:p-64 ui-grid-gap grid grid-cols-6"
  >
    <div class="col-span-full md:col-span-2">
      <div class="flex flex-row p-menu-row-snug">
        <img
          class="mr-24 -mt-16"
          src="/assets/ably_ui/core/images/ably-stack-f942defb86ef288e999c451c1f7e34fbbe59b7200d738c080e15a9f8716bbdad.svg"
          alt="Ably homepage"
        />
        <h2
          class="text-overline2 col-span-full font-medium uppercase tracking-widen-0.1"
        >
          The Ably Platform
        </h2>
      </div>
      <div class="grid grid-cols-4">
        <p class="text-p3 py-16 font-medium p-menu-row-snug col-span-3">
          Easily power any realtime experience in your application via a simple
          API that handles everything realtime.
        </p>
      </div>
      <ul
        class="grid gap-x-8 sm:gap-x-16 md:gap-x-24 xl:gap-x-32 sm:grid-cols-2 md:grid-cols-1"
      >
        <li class="p-menu-row-snug">
          <a href="/pub-sub-messaging" class="ui-footer-menu-row-link"
            >Pub/sub messaging</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/push-notifications" class="ui-footer-menu-row-link"
            >Push notifications</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/integrations" class="ui-footer-menu-row-link"
            >Third-party integrations</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/protocols" class="ui-footer-menu-row-link"
            >Multiple protocol messaging</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/hub" class="ui-footer-menu-row-link"
            >Streaming data sources</a
          >
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">Ably is for</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a href="/solutions/edtech" class="ui-footer-menu-row-link">EdTech</a>
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/solutions/automotive-logistics-and-mobility"
            class="ui-footer-menu-row-link"
            >Automotive, Logistics, &amp; Mobility</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/b2b-platforms" class="ui-footer-menu-row-link"
            >B2B Platforms</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/healthcare" class="ui-footer-menu-row-link"
            >Healthcare</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/hub" class="ui-footer-menu-row-link"
            >Streaming data sources</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/solutions/ecommerce-and-retail"
            class="ui-footer-menu-row-link"
            >eCommerce &amp; Retail</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/sports-and-media" class="ui-footer-menu-row-link"
            >Sports &amp; Media</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/gaming" class="ui-footer-menu-row-link">Gaming</a>
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/solutions/iot-and-connected-devices"
            class="ui-footer-menu-row-link"
            >IoT &amp; Connected Devices</a
          >
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">Developers</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a
            href="/documentation/quick-start-guide"
            class="ui-footer-menu-row-link"
            >Start in 5 minutes</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/documentation/" class="ui-footer-menu-row-link"
            >Documentation</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/tutorials" class="ui-footer-menu-row-link">Tutorials</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="https://changelog.ably.com/" class="ui-footer-menu-row-link"
            >Changelog</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/support" class="ui-footer-menu-row-link"
            >Support &amp; FAQs</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/download" class="ui-footer-menu-row-link">SDKs</a>
        </li>
        <li class="p-menu-row-snug flex items-center">
          <a
            class="pr-8 ui-footer-menu-row-link"
            href="https://status.ably.com/"
            >System status</a
          ><iframe
            class="w-24 h-24 mt-4"
            src="https://status.ably.io/embed/icon"
            allowtransparency="true"
            frameborder="0"
            scrolling="no"
          ></iframe>
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">WHY ABLY</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a href="/customers" class="ui-footer-menu-row-link"
            >Customer Stories</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/four-pillars-of-dependability"
            class="ui-footer-menu-row-link"
            >Four Pillars of Dependability</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/compare" class="ui-footer-menu-row-link"
            >Compare our tech</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/protocols" class="ui-footer-menu-row-link"
            >Multi protocol support</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/integrations" class="ui-footer-menu-row-link"
            >Third-party integrations</a
          >
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">ABOUT</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a href="/about" class="ui-footer-menu-row-link">About Ably</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/pricing" class="ui-footer-menu-row-link">Pricing</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/blog" class="ui-footer-menu-row-link">Blog</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/enterprise" class="ui-footer-menu-row-link">Enterprise</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/careers" class="ui-footer-menu-row-link">Careers</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/open-policy" class="ui-footer-menu-row-link"
            >Open protocol policy</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/contact" class="ui-footer-menu-row-link">Contact us</a>
        </li>
      </ul>
    </div>
  </div>
  <div class="max-w-screen-xl px-24 md:px-32 xl:px-64 mx-auto">
    <hr class="border-t-1 border-mid-grey" />
  </div>
  <div
    class="max-w-screen-xl mx-auto p-24 md:px-32 lg:p-32 xl:px-64 grid grid-cols-2"
  >
    <div
      class="flex flex-col flex-auto pb-40 ml-12 col-span-full md:col-span-1"
    >
      <div class="ui-footer-legals">
        <a href="/privacy" class="pr-24 ui-footer-link">Cookies</a
        ><a href="/legals" class="pr-24 ui-footer-link">Legals</a
        ><a href="/data-protection" class="ui-footer-link">Data Protection</a>
      </div>
      <div class="pt-32 flex items-center">
        <a
          class="h-24 pr-24 text-cool-black hover:text-icon-twitter"
          href="https://twitter.com/ablyrealtime"
          ><svg class="w-24 h-24">
            <use xlink:href="#sprite-twitter"></use></svg></a
        ><a
          class="h-24 pr-24 text-cool-black hover:text-icon-linkedin"
          href="https://www.linkedin.com/company/ably-realtime"
          ><svg class="w-24 h-24">
            <use xlink:href="#sprite-linkedin"></use></svg></a
        ><a
          class="h-24 pr-24 text-cool-black hover:text-icon-github"
          href="https://github.com/ably/"
          ><svg class="w-24 h-24"><use xlink:href="#sprite-github"></use></svg
        ></a>
        <div class="border-l-2 h-40"></div>
        <a
          class="flex flex-row pl-24 text-cool-black hover:text-icon-glassdoor"
          href="https://www.glassdoor.co.uk/Overview/Working-at-Ably-EI_IE2184188.11,15.htm"
          ><svg class="w-24 h-24">
            <use xlink:href="#sprite-glassdoor"></use></svg
        ></a>
        <div class="pl-16 text-menu3 font-light">
          <strong class="block font-medium">Were hiring!</strong
          ><a
            href="https://www.glassdoor.co.uk/Overview/Working-at-Ably-EI_IE2184188.11,15.htm"
            class="ui-footer-link"
            >Learn more</a
          >&nbsp;at Glass Door
        </div>
      </div>
    </div>
    <div class="md:text-right col-span-full md:col-span-1 ml-12">
      <img
        class="mr-24 w-96 h-96"
        src="/assets/ably_ui/core/images/rocket-list-2021-585749315fb7d34591caa21db10b0a2abc6bb38bbe32f9f350b5ae48a05c26dd.png"
        alt="Rocket List 2021"
      /><img
        class="w-96 h-96"
        src="/assets/ably_ui/core/images/flexible-companies-dc4c35aaec0e576c23df2d8c684bf64739a0df9ed6410f84ea7ba61245ca4efc.png"
        alt="Flexible Companies 2021"
      />
    </div>
  </div>
</footer>
`;

exports[`Footer Snapshot Test > view component > test case common to both #0`] = `
<footer class="bg-light-grey font-sans antialiased" data-id="footer">
  <div
    class="max-w-screen-xl mx-auto p-24 md:px-32 lg:p-32 xl:p-64 ui-grid-gap grid grid-cols-6"
  >
    <div class="col-span-full md:col-span-2">
      <div class="flex flex-row p-menu-row-snug">
        <img
          class="mr-24 -mt-16"
          src="/assets/ably_ui/core/images/ably-stack-f942defb86ef288e999c451c1f7e34fbbe59b7200d738c080e15a9f8716bbdad.svg"
          alt="Ably homepage"
        />
        <h2
          class="text-overline2 col-span-full font-medium uppercase tracking-widen-0.1"
        >
          The Ably Platform
        </h2>
      </div>
      <div class="grid grid-cols-4">
        <p class="text-p3 py-16 font-medium p-menu-row-snug col-span-3">
          Easily power any realtime experience in your application via a simple
          API that handles everything realtime.
        </p>
      </div>
      <ul
        class="grid gap-x-8 sm:gap-x-16 md:gap-x-24 xl:gap-x-32 sm:grid-cols-2 md:grid-cols-1"
      >
        <li class="p-menu-row-snug">
          <a href="/pub-sub-messaging" class="ui-footer-menu-row-link"
            >Pub/sub messaging</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/push-notifications" class="ui-footer-menu-row-link"
            >Push notifications</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/integrations" class="ui-footer-menu-row-link"
            >Third-party integrations</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/protocols" class="ui-footer-menu-row-link"
            >Multiple protocol messaging</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/hub" class="ui-footer-menu-row-link"
            >Streaming data sources</a
          >
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">Ably is for</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a href="/solutions/edtech" class="ui-footer-menu-row-link">EdTech</a>
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/solutions/automotive-logistics-and-mobility"
            class="ui-footer-menu-row-link"
            >Automotive, Logistics, &amp; Mobility</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/b2b-platforms" class="ui-footer-menu-row-link"
            >B2B Platforms</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/healthcare" class="ui-footer-menu-row-link"
            >Healthcare</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/hub" class="ui-footer-menu-row-link"
            >Streaming data sources</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/solutions/ecommerce-and-retail"
            class="ui-footer-menu-row-link"
            >eCommerce &amp; Retail</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/sports-and-media" class="ui-footer-menu-row-link"
            >Sports &amp; Media</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/solutions/gaming" class="ui-footer-menu-row-link">Gaming</a>
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/solutions/iot-and-connected-devices"
            class="ui-footer-menu-row-link"
            >IoT &amp; Connected Devices</a
          >
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">Developers</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a
            href="/documentation/quick-start-guide"
            class="ui-footer-menu-row-link"
            >Start in 5 minutes</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/documentation/" class="ui-footer-menu-row-link"
            >Documentation</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/tutorials" class="ui-footer-menu-row-link">Tutorials</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="https://changelog.ably.com/" class="ui-footer-menu-row-link"
            >Changelog</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/support" class="ui-footer-menu-row-link"
            >Support &amp; FAQs</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/download" class="ui-footer-menu-row-link">SDKs</a>
        </li>
        <li class="p-menu-row-snug flex items-center">
          <a
            class="pr-8 ui-footer-menu-row-link"
            href="https://status.ably.com/"
            >System status</a
          ><iframe
            class="w-24 h-24 mt-4"
            src="https://status.ably.io/embed/icon"
            allowtransparency="true"
            frameborder="0"
            scrolling="no"
          ></iframe>
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">WHY ABLY</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a href="/customers" class="ui-footer-menu-row-link"
            >Customer Stories</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a
            href="/four-pillars-of-dependability"
            class="ui-footer-menu-row-link"
            >Four Pillars of Dependability</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/compare" class="ui-footer-menu-row-link"
            >Compare our tech</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/protocols" class="ui-footer-menu-row-link"
            >Multi protocol support</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/integrations" class="ui-footer-menu-row-link"
            >Third-party integrations</a
          >
        </li>
      </ul>
    </div>
    <div class="col-span-full sm:col-span-3 md:col-span-1">
      <h2 class="ui-footer-col-title">ABOUT</h2>
      <ul>
        <li class="p-menu-row-snug">
          <a href="/about" class="ui-footer-menu-row-link">About Ably</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/pricing" class="ui-footer-menu-row-link">Pricing</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/blog" class="ui-footer-menu-row-link">Blog</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/enterprise" class="ui-footer-menu-row-link">Enterprise</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/careers" class="ui-footer-menu-row-link">Careers</a>
        </li>
        <li class="p-menu-row-snug">
          <a href="/open-policy" class="ui-footer-menu-row-link"
            >Open protocol policy</a
          >
        </li>
        <li class="p-menu-row-snug">
          <a href="/contact" class="ui-footer-menu-row-link">Contact us</a>
        </li>
      </ul>
    </div>
  </div>
  <div class="max-w-screen-xl px-24 md:px-32 xl:px-64 mx-auto">
    <hr class="border-t-1 border-mid-grey" />
  </div>
  <div
    class="max-w-screen-xl mx-auto p-24 md:px-32 lg:p-32 xl:px-64 grid grid-cols-2"
  >
    <div
      class="flex flex-col flex-auto pb-40 ml-12 col-span-full md:col-span-1"
    >
      <div class="ui-footer-legals">
        <a href="/privacy" class="pr-24 ui-footer-link">Cookies</a
        ><a href="/legals" class="pr-24 ui-footer-link">Legals</a
        ><a href="/data-protection" class="ui-footer-link">Data Protection</a>
      </div>
      <div class="pt-32 flex items-center">
        <a
          class="h-24 pr-24 text-cool-black hover:text-icon-twitter"
          href="https://twitter.com/ablyrealtime"
          ><svg class="w-24 h-24">
            <use xlink:href="#sprite-twitter"></use></svg></a
        ><a
          class="h-24 pr-24 text-cool-black hover:text-icon-linkedin"
          href="https://www.linkedin.com/company/ably-realtime"
          ><svg class="w-24 h-24">
            <use xlink:href="#sprite-linkedin"></use></svg></a
        ><a
          class="h-24 pr-24 text-cool-black hover:text-icon-github"
          href="https://github.com/ably/"
          ><svg class="w-24 h-24"><use xlink:href="#sprite-github"></use></svg
        ></a>
        <div class="border-l-2 h-40"></div>
        <a
          class="flex flex-row pl-24 text-cool-black hover:text-icon-glassdoor"
          href="https://www.glassdoor.co.uk/Overview/Working-at-Ably-EI_IE2184188.11,15.htm"
          ><svg class="w-24 h-24">
            <use xlink:href="#sprite-glassdoor"></use></svg
        ></a>
        <div class="pl-16 text-menu3 font-light">
          <strong class="block font-medium">Were hiring!</strong
          ><a
            href="https://www.glassdoor.co.uk/Overview/Working-at-Ably-EI_IE2184188.11,15.htm"
            class="ui-footer-link"
            >Learn more</a
          >&nbsp;at Glass Door
        </div>
      </div>
    </div>
    <div class="md:text-right col-span-full md:col-span-1 ml-12">
      <img
        class="mr-24 w-96 h-96"
        src="/assets/ably_ui/core/images/rocket-list-2021-585749315fb7d34591caa21db10b0a2abc6bb38bbe32f9f350b5ae48a05c26dd.png"
        alt="Rocket List 2021"
      /><img
        class="w-96 h-96"
        src="/assets/ably_ui/core/images/flexible-companies-dc4c35aaec0e576c23df2d8c684bf64739a0df9ed6410f84ea7ba61245ca4efc.png"
        alt="Flexible Companies 2021"
      />
    </div>
  </div>
</footer>
`;

exports[`ControlMobileDropdown > react > renders correctly by default #0`] = `
<button
  type="button"
  class="block ml-24 mr-0 px-0 py-16 hover:text-gui-hover focus:text-gui-focus focus:outline-none"
  data-id="meganav-control-mobile-dropdown"
  aria-expanded="false"
  aria-controls="meganav-mobile-dropdown"
>
  <svg
    class="h-24 w-24 transition-colors ui-icon-cool-black"
    data-id="meganav-control-mobile-dropdown-menu"
  >
    <use xlink:href="#sprite-menu"></use></svg
  ><svg
    class="h-24 w-24 transition-colors hidden ui-icon-cool-black"
    data-id="meganav-control-mobile-dropdown-close"
  >
    <use xlink:href="#sprite-close"></use>
  </svg>
</button>
`;

exports[`ControlMobileDropdown > react > renders correctly on click #0`] = `
<button
  type="button"
  class="block ml-24 mr-0 px-0 py-16 hover:text-gui-hover focus:text-gui-focus focus:outline-none"
  data-id="meganav-control-mobile-dropdown"
  aria-expanded="true"
  aria-controls="meganav-mobile-dropdown"
>
  <svg
    class="h-24 w-24 transition-colors ui-icon-cool-black hidden"
    data-id="meganav-control-mobile-dropdown-menu"
  >
    <use xlink:href="#sprite-menu"></use></svg
  ><svg
    class="h-24 w-24 transition-colors ui-icon-cool-black"
    data-id="meganav-control-mobile-dropdown-close"
  >
    <use xlink:href="#sprite-close"></use>
  </svg>
</button>
`;

exports[`ControlMobileDropdown > vw > renders correctly by default #0`] = `
<button
  type="button"
  class="block ml-24 mr-0 px-0 py-16 hover:text-gui-hover focus:text-gui-focus focus:outline-none"
  data-id="meganav-control-mobile-dropdown"
  aria-expanded="false"
  aria-controls="meganav-mobile-dropdown"
>
  <svg
    class="h-24 w-24 transition-colors ui-icon-cool-black"
    data-id="meganav-control-mobile-dropdown-menu"
  >
    <use xlink:href="#sprite-menu"></use></svg
  ><svg
    class="h-24 w-24 transition-colors hidden ui-icon-cool-black"
    data-id="meganav-control-mobile-dropdown-close"
  >
    <use xlink:href="#sprite-close"></use>
  </svg>
</button>
`;

exports[`ControlMobileDropdown > vw > renders correctly on click #0`] = `
<button
  type="button"
  class="block ml-24 mr-0 px-0 py-16 hover:text-gui-hover focus:text-gui-focus focus:outline-none"
  data-id="meganav-control-mobile-dropdown"
  aria-expanded="true"
  aria-controls="meganav-mobile-dropdown"
>
  <svg
    class="h-24 w-24 transition-colors ui-icon-cool-black hidden"
    data-id="meganav-control-mobile-dropdown-menu"
  >
    <use xlink:href="#sprite-menu"></use></svg
  ><svg
    class="h-24 w-24 transition-colors ui-icon-cool-black"
    data-id="meganav-control-mobile-dropdown-close"
  >
    <use xlink:href="#sprite-close"></use>
  </svg>
</button>
`;
